retrigger/triggerhandler.py

```diff
@@ -368,6 +396,9 @@ async def check_triggers(self, message: discord.Message, edit: bool) -> None:
             if trigger.ocr_search and ALLOW_OCR:
                 content += await self.get_image_text(message)
 
+            if trigger.embeds and len(message.embeds) > 0:
+                content += " " + " ".join(str(embed.to_dict()) for embed in message.embeds)
```

Comment:
I am not a fan of just passing the string of the dictionary to the filter check. It can cause unnecessary problems for user to utilize this feature having to build their regex on embeds around a python dictionary object. It would be better to itemize out the important visible parts of the embed objects like title, description, url, etc. We can probably ignore things like the proxy_icon_url since those are discord cdn urls anyways and could give the false impression if someone has poor regex to filter discord links.

Reply:
Sure thing, I will look at filtering this down, this was purely for the sake of not "over-complicating" it and adding a big chunk of logic. I will get some data on multiple different types of URL embeds and what they contain because I know that Discord treats things like Twitter and Youtube "special" adding specific fields for these platforms.

Reply:
Okay, trying a few sources such as
- Twitter
- Youtube
- Github
- Reddit
- Generic URL

I think the below is what may be good to pick out:
- Author
  - Name
  - URL
- Title
- Description
- Type
- Provider
  - Name
- URL
- Image
  - URL
- Video
  - URL

Reply:
I don't believe  the `Type` is relevant, that's an internal discord embed type. It's only ever going to be one of [these.](https://discord.com/developers/docs/resources/channel#embed-object-embed-types) If you're filtering by any of those types of embeds you might as well just toggle the ability for users to embed links in the channel entirely.

You're also missing the `fields`, `footer`, `thumbnail`, and `image` sections of an embed which could contain content you may want filtered. A embed could link to one page but contain an image or thumbnail from another webpage you absolutely don't want shared for example.

 You don't need to investigate the different types of links to see this it's [documented here](https://discord.com/developers/docs/resources/channel#embed-object-embed-structurel) for all embed structures.

To keep the code clean I recommend building a method `def convert_embed_to_string(self, embed: discord.Embed) -> str` or something along those lines that pulls the necessary attributes of the embed into a structured string format that is not just the dictionary. I expect the end string would look something like ```
Author: name
url
Title: title
Description: description
Provider: provider
URL: url
Field 1: field 1 name
field 1 value
Field 2: field 2 name
field 2 value
Footer: footer
```
In order to not complicate regex searches it may be even better to not provide the descriptions for each attribute and just provide the string content of each separated by a new line.

Reply:
You are right that the fields are documented, but unsurprisingly Discord does not treat all links equally in how it fills the embed fields for different sites. That was the reason for testing a multitude of sites.

The `Type` isn't static for links, it isn't `link` for all links. In saying that you are right in that it is probably not needed, the only types I have seen in testing are `rich`, `video` and `link` all of which are pretty useless by themselves and there are probably other, better ways to filter these types of links.

To keep to the original scope of just covering auto-generated URL embeds below are some comments on fields based on my findings from testing with the various URL types above:
*Fields*
- Not used for nearly all URL embeds, only used for Twitter in which case it only contains the like/dislike count
*Footer*
- Again only used for Twitter, the only thing it contains is the Twitter logo and timestamp
*Thumbnail*
- Might actually be useful to filter on this, the thumbnail is an image from the page for "Generic URL's" for most other things it is pretty useless though. MIGHT provide a static image for Youtube but it is not clear how thumbnail is represented visually here.
*Image*
- I did remove this but you are right in that it should be added, in Twitter it contains an attached image URL and in Imgur tests it contains the image itself
*Video*
- Same as above, I removed it but with Imgur this contains the actual image URL even when a gallery link is used.

My view is that it will be best to leave the descriptions. This is because say for example you want to trigger on posts from a certain Twitter user that may have a very generic dictionary-based name. If you don't give the option to filter by `\bAuthorName: GenericUsername\b" then you could end up with it triggering on every single use of the word which matches the generic username.

Reply:
It should be noted that while you've tested a handful of sites, and it looks like Discord does some custom backend work for things like twitter, technically all embed fields are possible to be manipulated through a variety of means. That is to say a user can host a webpage that contains all the necessary information to build a [completely custom](https://www.reddit.com/r/discordapp/comments/82p8i6/a_basic_tutorial_on_how_to_get_the_most_out_of/) embed so all fields should be accounted for. As mentioned the type is irrelevant for the searching we just want to search things that can be customized in some way. 

The string should also avoid having something like `Author: None` in situations where that embed field does not exist. I also want to point out that if I wanted to filter something like twitter from a specific user it would be trivial to just filter the twitter URL containing tweets from said user through already existing means.

Reply:
That is interesting! I wasn't aware of that avenue of manipulating the Embed data, I thought we were accidentally creeping into bot or self-bot created custom embeds setting attributes. I will definitely expand that data set then 😄

Reply:
New function has been created for removing unwanted embed fields and formatting everything as a sensible string. During testing I added a print, below is how a message containing a Twitter url embed looks afterwards:
```

redbot   | Url: https://twitter.com/KaiyokoStar/status/1664559787982352384?s=20
redbot   | https://twitter.com/KaiyokoStar/status/1654427905907204097?s=20
redbot   | FooterText: Twitter
redbot   | FooterIcon_url: https://abs.twimg.com/icons/apple-touch-icon-192x192.png
redbot   | ImageUrl: https://pbs.twimg.com/media/FvW1ZDFWAAAh_N3.jpg
redbot   | AuthorName: Kaiyoko 💫 (@KaiyokoStar)
redbot   | AuthorIcon_url: https://pbs.twimg.com/profile_images/1634341028315496449/zRm5_IHw_400x400.jpg
redbot   | AuthorUrl: https://twitter.com/KaiyokoStar
redbot   | Fields: [{'name': 'Likes', 'value': '399', 'inline': True}]
redbot   | Description: Fashion Report Week 275 - Full Details #ffxiv #ff14 #ファッションチェック
redbot   | Url: https://twitter.com/KaiyokoStar/status/1654427905907204097?s=20
```

Reply:
Got another revision still to go here to parse out the children of `Fields`

Reply:
With that done here is an example with Embed field `Fields` parsed out. Given all fields are name-value rather than using Field1, Field2... I opted with appending the `name` value to Fields then using the `value` as the direct value:
```
redbot   | https://twitter.com/KaiyokoStar/status/1626493822220771328?s=20
redbot   | FooterText: Twitter
redbot   | FooterIcon_url: https://abs.twimg.com/icons/apple-touch-icon-192x192.png
redbot   | ImageUrl: https://pbs.twimg.com/media/FpJ3akhXwAU2hGD.jpg
redbot   | AuthorName: Kaiyoko 💫 (@KaiyokoStar)
redbot   | AuthorIcon_url: https://pbs.twimg.com/profile_images/1634341028315496449/zRm5_IHw_400x400.jpg
redbot   | AuthorUrl: https://twitter.com/KaiyokoStar
redbot   | FieldsLikes: 168
redbot   | Description: Fashion Report Week 264 - Full Details #ffxiv #ff14 #ファッションチェック
redbot   | Url: https://twitter.com/KaiyokoStar/status/1626493822220771328?s=20
```